lib/Twig/Cache/Filesystem.php

```diff
@@ -29,11 +29,11 @@ public function __construct($directory)
     /**
      * {@inheritdoc}
      */
-    public function generateKey($className, $prefix)
+    public function generateKey($name, $className)
     {
-        $class = substr($className, strlen($prefix));
+        $len = strlen($className);
 
-        return $this->directory.'/'.$class[0].'/'.$class[1].'/'.$class.'.php';
+        return $this->directory.'/'.$className[$len - 1].'/'.$className[$len - 2].'/'.$className.'.php';
```

Comment:
The class name might contain the index with a `_x` suffix. In this case the directory structure is not randomized evenly anymore. A simple solution would be to just change ´getTemplateClass` to add the index between prefix and hash.
But this still ties to Cache implementation to the class generation, i.e. the cache is basically aware how the class name is generated.